### PR TITLE
Bugfix 1286 develop pb2nc

### PR DESCRIPTION
**Pull request by hsoh-u**
Pull request for development branch
See #1286 for details and #1300 for PR for master branch
Another pull request for develop branch

### CODE REVIEW

File: met/src/tools/other/pb2nc/pb2nc.cc [hunk 1]
```diff
@@ -1513,38 +1526,45 @@ void process_pbfile(int i_pb) {
...
-               // Increment the current and total observations counts
-               n_file_obs++;
-               n_total_obs++;
-
-               // Increment the number of obs counter for this header
-               n_hdr_obs++;
-            }
-         } // end for i
+         if (n_derive_gc > bufr_derive_cfgs.size()) {
+            mlog << Warning << "\n" << method_name << " -> "
```
Comment (JohnHalleyGotway) [inline met/src/tools/other/pb2nc/pb2nc.cc]: Recommend changing "Warning" to "Debug(3)".  Not requesting any derivations shouldn't be a warning, just an informational message.

Comment (JohnHalleyGotway) [reply met/src/tools/other/pb2nc/pb2nc.cc]: I actually don't understand this logic.  n_derive_gc = 6.  If I configure pb2nc to only derive DPT, for example, what is bufr_derive_cfgs.size()?

**Review by JohnHalleyGotway — COMMENTED** [review]
I tested in kiowa:/d1/projects/MET/MET_pull_requests/pull_request_1301.  I compiled the bug_1286_develop_pb2nc_mem_leak branch locally, ran unit_pb2nc.xml, and confirmed that all the tests run.  It is not really possible to directly confirm that these changes fix the problem that originally occurred on WCOSS.  But I inspected the code and it generally looks good.  I am waiting for more explanation of pb2nc.cc line 1530.

**Comment by hsoh-u** [comment]
before:
static int bufr_derive_code[n_derive_gc];
static StringArray prepbufr_derive_vars;

after:
static StringArray prepbufr_derive_vars;
static vector<derive_var_cfg> bufr_derive_cfgs;

Step 1.  at initialize() method:
prepbufr_derive_vars is initialized. It has 8 (> n_derive_gc) with additional D_PBL and D_CAPE.

Step 2.  at process_pbfile_metadata() method:
It checks which variables are available and which variables are requested. And set bufr_derive_cfgs (see code block below). The size of bufr_derive_cfgs is 0 (empty If all PTQZUV variables are not available) OR the same as the size of prepbufr_derive_vars (i.e. > n_derive_gc).

The derived variable is requested, the bufr_derive_cfgs[vIdx].var_index contains a valid variable index (not negative integer). The negative integer at bufr_derive_cfgs[vIdx].var_index is filtered by keep_bufr_obs_index().

=== code block ===

   bufr_derive_cfgs.clear();
   if (has_prepbufr_vars) {
      for (int vIdx=0; vIdx< prepbufr_derive_vars.n_elements(); vIdx++) {
         tmp_var_name = prepbufr_derive_vars[vIdx].c_str();
         bufr_derive_cfgs.push_back(derive_var_cfg(tmp_var_name));
         if (do_all_vars || bufr_target_variables.has(tmp_var_name)) {
            // Set the variable index if requested
            bufr_obs_name_arr.add(tmp_var_name);
            bufr_derive_cfgs[vIdx].var_index = bufr_var_index;
            bufr_var_index++;
         }
      }
   }

**Comment by JohnHalleyGotway** [comment]
Thanks for the explanation Howard.

**Pull request merged**